phoenix-core/src/test/java/org/apache/phoenix/index/PrepareIndexMutationsForRebuildTest.java

```diff
@@ -0,0 +1,727 @@
+import org.apache.phoenix.hbase.index.IndexRegionObserver;
+import org.apache.phoenix.hbase.index.util.GenericKeyValueBuilder;
+import org.apache.phoenix.jdbc.PhoenixConnection;
+import org.apache.phoenix.query.BaseConnectionlessQueryTest;
+import org.apache.phoenix.query.QueryConstants;
+import org.apache.phoenix.schema.PName;
+import org.apache.phoenix.schema.PTable;
+import org.apache.phoenix.schema.PTableKey;
+import org.apache.phoenix.util.SchemaUtil;
+import org.junit.Assert;
+import org.junit.Test;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.List;
+
+import static org.junit.Assert.assertEquals;
+
+public class PrepareIndexMutationsForRebuildTest extends BaseConnectionlessQueryTest {
+    private static String ROW_KEY = "k1";
+    private static String TABLE_NAME = "dataTable";
+    private static String INDEX_NAME = "idx";
+    private static final Logger LOGGER = LoggerFactory.getLogger(PrepareIndexMutationsForRebuildTest.class);
+
+    class SetupInfo{
+        public IndexMaintainer indexMaintainer;
+        public PTable pDataTable;
+    }
+
+    private SetupInfo setupIndexMaintainer(String tableName,
+                                                 String indexName,
+                                                 String columns,
+                                                 String indexColumns,
+                                                 String pk,
+                                                 String includeColumns) throws Exception{
+        Connection conn = DriverManager.getConnection(getUrl());
+
+        String fullTableName = SchemaUtil.getTableName(SchemaUtil.normalizeIdentifier(""),SchemaUtil.normalizeIdentifier(tableName));
+        String fullIndexName = SchemaUtil.getTableName(SchemaUtil.normalizeIdentifier(""),SchemaUtil.normalizeIdentifier(indexName));
+
+        String str1 = String.format("CREATE TABLE %1$s (%2$s CONSTRAINT pk PRIMARY KEY (%3$s)) COLUMN_ENCODED_BYTES=0",
+                fullTableName,
+                columns,
+                pk);
+        conn.createStatement().execute(str1);
+
+        String str2 = String.format("CREATE INDEX %1$s ON %2$s (%3$s)",
+                fullIndexName,
+                fullTableName,
+                indexColumns);
+        if (!includeColumns.isEmpty())
+            str2 += " INCLUDE (" + includeColumns + ")";
+        conn.createStatement().execute(str2);
+
+        PhoenixConnection pconn = conn.unwrap(PhoenixConnection.class);
+        PTable pDataTable = pconn.getTable(new PTableKey(pconn.getTenantId(), fullTableName));
+        ImmutableBytesWritable ptr = new ImmutableBytesWritable();
+        pDataTable.getIndexMaintainers(ptr, pconn);
+        List<IndexMaintainer> indexMaintainers = IndexMaintainer.deserialize(ptr, GenericKeyValueBuilder.INSTANCE, true);
```

Comment:
I think it will be more readable and cleaner if you get the index maintainer object from the index table using getIndexMaintainer()